### PR TITLE
filter valid_systems on gpu_vendor extras

### LINKED ISSUE #34: add support for specifying GPU vendor in configuration
with feature `cuda` we can more explicitly filter on devices that support cuda (required for tests that use a cuda module on a gpu).

later on, we more easily add `rocm` and `oneapi` when AMD and Intel GPU modules become available in EESSI.

### LINKED ISSUE #40: reframe.log and reframe.out contents get overwritten
The contents of the reframe.log file and reframe.out file get overwritten with every run by reframe. 

Currently keeping the logs and outputs of different runs by changing the names of the files manually.

### PR DESCRIPTION
**Pull request by smoors**
fixes #34 
fixes #40 

this PR changes `filter_valid_systems_by_device_type` in `hooks.py` to explicitly require an NVIDIA GPU

from
```python
        if is_cuda_module and required_device_type == DEVICES['GPU']:
            # CUDA modules and when using a GPU require partitions with 'gpu' feature
            valid_systems = f'+{FEATURES["GPU"]}'
```
to
```python
    if is_cuda_module and required_device_type == DEVICES['GPU']:
        # CUDA modules and when using a GPU require partitions with 'gpu' feature and
        # 'gpu_vendor=nvidia' extras
        valid_systems = f'+{FEATURES["GPU"]} %gpu_vendor={GPU_VENDORS["NVIDIA"]}'
```

### CODE REVIEW

File: config/surf_snellius.py [hunk 1]
```diff
@@ -59,20 +59,23 @@
                     'features': [
                         'gpu',
                     ],
+                    'extras': {
+                        'gpu_vendor': 'nvidia',
```
Comment (boegel) [inline config/surf_snellius.py:63]: Use `GPU_VENDORS` constant here too? same for `FEATURES`

File: config/vsc_hortense.py [hunk 2]
```diff
@@ -124,6 +124,9 @@ def command(self, job):
                     'features': [
                         'gpu',
                     ],
+                    'extras': {
+                        'gpu_vendor': 'nvidia',
```
Comment (boegel) [inline config/vsc_hortense.py:128]: Use `GPU_VENDORS` constant here too? same for `FEATURES`

File: config/settings_example.py [hunk 3]
```diff
@@ -58,7 +58,13 @@
                             'num_devices': 4,
                         }
                     ],
-                    'features': [FEATURES['CPU'], FEATURES['GPU']],
+                    'features': [
+                        FEATURES['CPU'],
+                        FEATURES['GPU'],
+                    ],
+                    'extras': {
+                        'gpu_vendor': GPU_VENDORS['NVIDIA'],
```
Comment (boegel) [inline config/settings_example.py:66]: introduce `GPU_VENDOR` + `NVIDIA` constants?

```suggestion
                        GPU_VENDOR: GPU_VENDORS[NVIDIA],

File: eessi/testsuite/hooks.py [hunk 4]
```diff
@@ -180,25 +180,28 @@ def filter_valid_systems_by_device_type(test: rfm.RegressionTest, required_devic
...
+
+    is_cuda_module = is_cuda_required_module(test.module_name)
 
-        if is_cuda_module and required_device_type == DEVICES['GPU']:
-            # CUDA modules and when using a GPU require partitions with 'gpu' feature
-            valid_systems = f'+{FEATURES["GPU"]}'
+    if is_cuda_module and required_device_type == DEVICES['GPU']:
+        # CUDA modules and when using a GPU require partitions with 'gpu' feature and
+        # 'gpu_vendor=nvidia' extras
+        valid_systems = f'+{FEATURES["GPU"]} %gpu_vendor={GPU_VENDORS["NVIDIA"]}'
```
Comment (boegel) [inline eessi/testsuite/hooks.py:192]: use `GPU_VENDOR` constant:

```suggestion
        valid_systems = f'+{FEATURES["GPU"]} %{GPU_VENDOR}={GPU_VENDORS["NVIDIA"]}'

File: README.md [hunk 5]
```diff
@@ -46,16 +48,15 @@ reframe \
 ## Configuring GPU/non-GPU partitions in your site config file:
 
 - running GPU jobs in GPU nodes
-    - add feature `gpu` to the GPU partitions
+    - add `'features': [gpu]` to the GPU partitions
+    - add `'extras': {'gpu_vendor': 'nvidia'}` to the GPU partitions (or
```
Comment (boegel) [inline README.md:52]: use `GPU_VENDOR` and `GPU_VENDORS` constants here too

File: config/settings_example.py [hunk 6]
```diff
@@ -3,14 +3,14 @@
 """
 from os import environ
 
-from eessi.testsuite.constants import DEVICES, FEATURES
+from eessi.testsuite.constants import DEVICES, FEATURES, GPU_VENDORS
```
Comment (boegel) [inline config/settings_example.py:6]:
```
from eessi.testsuite.constants import *
```

**Review by boegel — DISMISSED** [review]

File: config/vsc_hortense.py [hunk 7]
```diff
@@ -65,7 +68,7 @@ def command(self, job):
                         'arch': 'zen2',
                     },
                     'features': [
-                        'cpu',
+                        FEATURES[CPU],
```
Comment (casparvl) [inline config/vsc_hortense.py:71]: Wait... this should be `FEATURES['CPU']`, right? I mean, indexing the map happens with a string `'CPU'` not a python object `CPU`

Comment (casparvl) [reply config/vsc_hortense.py:71]: Oh, never mind, I just checked your `constants.py` and see what you did there... `CPU='CPU'`. It does _look_ very readable like this. The only concern I have is: it changes quite short object names in de global namespace. Is that wise?

Comment (smoors) [reply config/vsc_hortense.py:71]: hm, not sure if it really is a problem.
if you think it is, we could make the constants longer and do away with the dictionaries, e.g.
```
FEATURE_CPU = 'cpu'
FEATURE_GPU = 'gpu'
GPU_VENDOR_NVIDIA = 'nvidia'
GPU_VENDOR_AMD = 'amd'
```

the proper way is with namespacing, but that becomes a lot less readable

Comment (casparvl) [reply config/vsc_hortense.py:71]: Let's keep it like you proposed originally for now, it is indeed the cleanest way. We'll find out if it comes back to haunt us some day ;-)

**Review by casparvl — APPROVED** [review]
Looks good to me. Testing it right now with the GROMACS test. That was send to the correct partition (i.e. if `nb_impl=gpu`, it is submitted to a GPU capable partition, that has nvidia as `extra`). The test itself is still waiting in the queue. Though there shouldn't be any changes that affect how the test _runs_, I'll await it's succesful completion before mergin this PR, just to be sure.

**Pull request merged**